WalletWasabi/WabiSabi/Client/SmartRequestNode.cs

```diff
@@ -134,37 +112,46 @@ private IEnumerable<long> AddExtraCredential(IEnumerable<long> valuesToRequest,
 				return nonZeroValues;
 			}
 
-			var missing = presentedCredentials.Sum(cr => cr.Value) - valuesToRequest.Sum();
+			var missing = sum - valuesToRequest.Sum();
 
 			if (missing > 0)
 			{
 				nonZeroValues = nonZeroValues.Append(missing);
 			}
 
+			// Note that this does not include the implied zero credentials
+			// which are unconditionally requested.
 			var additionalZeros = ProtocolConstants.CredentialNumber - nonZeroValues.Count();
 
 			return nonZeroValues.Concat(Enumerable.Repeat(0L, additionalZeros));
 		}
 
-		private IEnumerable<long> AddExtraCredential(IEnumerable<long> valuesToRequest, long sum)
+		private (IEnumerable<Credential>, IEnumerable<Credential>) SeparateExtraCredentials(IEnumerable<Credential> issuedCredentials, IEnumerable<long> requiredValues)
 		{
-			var nonZeroValues = valuesToRequest.Where(v => v > 0);
+			var taggedCredentials = TagExtraCredentials(issuedCredentials, requiredValues).ToImmutableArray();
 
-			if (nonZeroValues.Count() == ProtocolConstants.CredentialNumber)
-			{
-				return nonZeroValues;
-			}
+			return (
+				taggedCredentials.Where(x => !x.IsExtra).Select(x => x.Credential),
+				taggedCredentials.Where(x => x.IsExtra).Select(x => x.Credential)
+			);
+		}
 
-			var missing = sum - valuesToRequest.Sum();
+		private IEnumerable<(bool IsExtra, Credential Credential)> TagExtraCredentials(IEnumerable<Credential> issuedCredentials, IEnumerable<long> requiredValues)
+		{
+			var requiredEnumerator = requiredValues.GetEnumerator();
+			requiredEnumerator.MoveNext();
```

Comment:
This method is wrong in many ways. Replace with something like this (untested):

```c#
			return issuedCredentials.Select(c => (!requiredValues.Contains(c.Amount), c));

```

Reply:
ok, it is not the same. I see.

Reply:
just for clarification. enumerators has to be disposed and it is not okay to assume it always has at least one value.

Reply:
Maybe this one could work (untested again):

```c#
		private IEnumerable<(bool IsExtra, Credential Credential)> TagExtraCredentials(IEnumerable<Credential> issuedCredentials, IEnumerable<long> requiredValues)
		{
			using var requiredEnumerator = requiredValues.OrderBy(x => x).GetEnumerator();
			using var issuedEnumerator = issuedCredentials.OrderBy(x => x.Amount).GetEnumerator();

			while (requiredEnumerator.MoveNext())
			{
				var required = requiredEnumerator.Current;
				while (issuedEnumerator.MoveNext())
				{
					var issued = issuedEnumerator.Current;
					var isRequired = issued.Amount == required;
					yield return (!isRequired, issued)
					if (isRequired)
						break;
				}
			}
		}


```

Reply:
ack re disposal.

i had a similar control flow structure to this before and frankly, i hate it... the issued ones will always be strictly a superset of the requested ones so I much prefer having just a `foreach` for the issued ones, and "skipping over" the unrequested ones.

I think if the first `MoveNext()` fails (`requiredValues is empty`) we can just bypass it and `return Enumerable.Zip( Enumreable.Repeat(true), issuedCredentials)`.

Reply:
ok, i give up, going to go with your suggested structure even though i find it much less intuitive than nesting the loops the other way around, but it's just too ugly once the fact that the required value enumerator can be exhausted